mlx/backend/metal/kernels/quantized.h

```diff
@@ -442,25 +628,34 @@ METAL_FUNC void qmv_fast_impl(
     uint3 tid [[threadgroup_position_in_grid]],
     uint simd_gid [[simdgroup_index_in_threadgroup]],
     uint simd_lid [[thread_index_in_simdgroup]]) {
-  constexpr int packs_per_thread = bits > 2 ? 2 : 1;
+  constexpr int power_of_2_bits = (bits & (bits - 1)) == 0;
+  constexpr int packs_per_thread = bits == 2 ? 1 : 2;
   constexpr int num_simdgroups = 2;
   constexpr int results_per_simdgroup = 4;
-  constexpr int pack_factor = 32 / bits;
+  constexpr int pack_factor = bits == 3 ? 8 : bits == 6 ? 4 : 32 / bits;
+  constexpr int bytes_per_pack = power_of_2_bits ? 1 : 3;
   constexpr int values_per_thread = pack_factor * packs_per_thread;
   constexpr int block_size = values_per_thread * SIMD_SIZE;
   constexpr int scale_step_per_thread = group_size / values_per_thread;
 
+  // When bits is a power of two, read 1 uint32_t at a time
+  // When bits is 3 or 6, read 3 uint8_ts at a time
+  using W_T =
+      typename ConditionalType<power_of_2_bits, uint32_t, uint8_t>::type;
```

Comment:
Although I love the conditional type, I think we could just always use uint8 can't we?

Just change the bytes per pack to 4. Regardless the loading functions read in whatever type is convenient.

Reply:
I tried this and it was a bit slower for some reason, hence the conditional type stuff. Maybe I can try tuning that way a bit more, it would be nicer.

Reply:
Huh interesting. We can merge it as is and investigate later. This basically should result in exactly the same code for 2, 4 and 8 bits.

Reply:
Sounds good to me. For prompt processing it didn't make a difference so it's all uint8 in the `qmm`s.